activerecord/test/cases/adapter_test.rb

```diff
@@ -329,7 +329,7 @@ def test_select_methods_passing_a_association_relation
       author = Author.create!(name: "john")
       Post.create!(author: author, title: "foo", body: "bar")
       query = author.posts.where(title: "foo").select(:title)
-      assert_equal({ "title" => "foo" }, @connection.select_one(query))
+      assert_equal({ "title" => "foo" }, @connection.select_one(query).to_h)
```

Comment:
Does this PR mean change of public API?

Reply:
Yes, as [described above](https://github.com/rails/rails/pull/33312/files#r251207542) it changes the API. Even if `PG::Tuple` implements some `Hash` like methods, it never equals to a `Hash`, so `to_h` is required for comparison. 

@rafaelfranca called `ActiveRecord::Result` a ["public API by mistake"](https://github.com/rails/rails/pull/23404#issuecomment-178225051). Although we have several scripts at work making use of this API, it never felt like a public API, since it is halfway between the model query interface and the `raw_connection`. It also changed several times in the past, due to the needs within AR.

If public API change is a concern, it is possible to exclude `select_foo` methods from this change and let only `exec_foo` methods return the specialized `PostgreSQL::Result`. I changed `select_foo` for consistency, but they are not used by model query methods and are therefore not so much relevant for performance.